modtools/change-build-menu.lua

```diff
@@ -0,0 +1,560 @@
+        Remove the workshop or furnace with the ID <ID> from <CATEGORY>.
+        
+        <CATEGORY> and <ID> may have the same values as for the "add" option.
+    
+    change-build-menu revert <ID> <CATEGORY>
+        Revert an earlier remove or add operation. It is NOT safe to "remove"
+        and "add"ed building or vice versa, use this option to reverse any
+        changes you no longer want/need.
+    
+* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
+
+Module Usage:
+
+    To use this script as a module put the following somewhere in your own script:
+    
+        local buildmenu = reqscript "change-build-menu"
+    
+    Then you can call the functions documented here like so:
+    
+        -- Example: Remove the carpenters workshop.
+        buildmenu.ChangeBuilding("CARPENTERS", "WORKSHOPS", false)
+    
+    Note that to allow any of your changes to take effect you need to start the
+    ticker. See the "Command Usage" section.
+    
+Global Functions:
+    
+    function GetWShopID(btype, bsubtype, bcustom)
+        GetWShopID returns a workshop or furnace's string ID based on it's
+        numeric ID triplet. This string ID *should* match what is expected
+        by eventful for hardcoded buildings.
+    
+    function GetWShopType(id)
+        GetWShopIDs returns a workshop or furnace's ID numbers as a table.
+        The passed in ID should be the building's string identifier, it makes
+        no difference if it is a custom building or a hardcoded one.
+        The return table is structured like so: `{type, subtype, custom}`
+    
+    function IsEntityPermitted(id)
+        IsEntityPermitted returns true if DF would normally allow you to build
+        a workshop or furnace. Use this if you want to change a building, but
+        only if it is permitted in the current entity. You do not need to
+        specify an entity, the current fortress race is used.
+    
+    ChangeBuilding(id, category, [add, [key]])
+    ChangeBuildingAdv(typ, subtyp, custom, category, [add, [key]])
+        These two functions apply changes to the build sidebar menus. If "add"
+        is true then the building is added to the specified category, else it
+        is removed. When adding you may specify "key", a string DF hotkey ID.
+        
+        The first version of this function takes a workshop or furnace ID as a
+        string, the second takes a numeric ID triplet (which can specify any
+        building, not just workshops or furnaces).
+    
+    function RevertBuildingChanges(id, category)
+    function RevertBuildingChangesAdv(typ, subtyp, custom, category)
+        These two functions revert changes made by "ChangeBuilding" and
+        "ChangeBuildingAdv". Like those two functions there are two versions,
+        a simple one that takes a string ID and one that takes a numeric ID
+        triplet.
+]==]
+end
+
+if not dfhack.isWorldLoaded() then
+    dfhack.color(RED)

```

Comment:
Looking at it I can only assume I was sleep deprived at the time, since printerr would do exactly what I had intended... Like I said, I can't remember what I was thinking (was I thinking?). I'll change it.